pkg/api/api.go

```diff
@@ -0,0 +1,304 @@
+// Prometheus uses this when logging queries to QueryLogger, but Cortex doesn't call engine.SetQueryLogger to set one.
+//
+// Can be removed when (if) https://github.com/prometheus/prometheus/pull/6840 is merged.
+func fakeRemoteAddr(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.RemoteAddr == "" {
+			r.RemoteAddr = "127.0.0.1:8888"
+		}
+		handler.ServeHTTP(w, r)
+	})
+}
+
+// RegisterAlertmanager registers endpoints associated with the alertmanager. It will only
+// serve endpoints using the legacy http-prefix if it is not run as a single binary.
+func (a *API) RegisterAlertmanager(am *alertmanager.MultitenantAlertmanager, target bool) {
+	// Ensure this route is registered before the prefixed AM route
+	a.registerRoute("/multitenant-alertmanager/status", am.GetStatusHandler(), false)
+
+	// UI components lead to a large number of routes to support, utilize a path prefix instead
+	a.server.HTTP.PathPrefix(a.cfg.AlertmanagerHTTPPrefix).Handler(a.authMiddleware.Wrap(am))
+	level.Debug(a.logger).Log("msg", "api: registering alertmanager", "path_prefix", a.cfg.AlertmanagerHTTPPrefix)
+
+	// If the target is Alertmanager, enable the legacy behaviour. Otherwise only enable
+	// the component routed API.
+	if target {
+		a.registerRoute("/status", am.GetStatusHandler(), false)
+		a.server.HTTP.PathPrefix(a.cfg.LegacyHTTPPrefix).Handler(a.authMiddleware.Wrap(am))
+	}
+}
+
+// RegisterAPI registers the standard endpoints associated with a running Cortex.
+func (a *API) RegisterAPI(cfg interface{}) {
+	a.registerRoute("/config", configHandler(cfg), false)
+	a.registerRoute("/", http.HandlerFunc(indexHandler), false)
+}
+
+// RegisterDistributor registers the endpoints associated with the distributor.
+func (a *API) RegisterDistributor(d *distributor.Distributor, pushConfig distributor.Config) {
+	a.registerRoute("/api/v1/push", push.Handler(pushConfig, d.Push), true)
+	a.registerRoute("/distributor/all_user_stats", http.HandlerFunc(d.AllUserStatsHandler), false)
+	a.registerRoute("/distributor/ha-tracker", d.Replicas, false)
+
+	// Legacy Routes
+	a.registerRoute(a.cfg.LegacyHTTPPrefix+"/push", push.Handler(pushConfig, d.Push), true)
+	a.registerRoute("/all_user_stats", http.HandlerFunc(d.AllUserStatsHandler), false)
+	a.registerRoute("/ha-tracker", d.Replicas, false)
+}
+
+// RegisterIngester registers the ingesters HTTP and GRPC service
+func (a *API) RegisterIngester(i *ingester.Ingester, pushConfig distributor.Config) {
+	client.RegisterIngesterServer(a.server.GRPC, i)
+	grpc_health_v1.RegisterHealthServer(a.server.GRPC, i)
+
+	a.registerRoute("/ingester/flush", http.HandlerFunc(i.FlushHandler), false)
+	a.registerRoute("/ingester/shutdown", http.HandlerFunc(i.ShutdownHandler), false)
+	a.registerRoute("/ingester/push", push.Handler(pushConfig, i.Push), true) // For testing and debugging.
```

Comment:
That would work for me. I was also thinking it could live under `/ingester/debug/push`. Since it's only for debugging/testing. It might be nice to make that explicit in the path so it's clear to new users not to use it.

Reply:
Making it explicit could be a good idea 👍